src/java.base/share/classes/jdk/internal/classfile/impl/WritableField.java

```diff
@@ -0,0 +1,90 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile.impl;
+
+import java.lang.classfile.constantpool.ConstantPoolBuilder;
+import java.lang.classfile.constantpool.Utf8Entry;
+import java.util.Arrays;
+
+import static java.lang.classfile.ClassFile.ACC_STATIC;
+import static java.lang.classfile.ClassFile.ACC_STRICT;
+
+/**
+ * An interface to obtain field properties for direct class builders.
+ * Required to filter strict instance fields for stack map generation.
+ * Public for benchmark access.
+ */
+public sealed interface WritableField extends Util.Writable
+        permits FieldImpl, DirectFieldBuilder {
+    Utf8Entry fieldName();
+    Utf8Entry fieldType();
+    int fieldFlags();
+
+    static WritableField.UnsetField[] filterStrictInstanceFields(ConstantPoolBuilder cpb, WritableField[] array, int count) {
+        // assume there's no toctou for trusted incoming array
+        int size = 0;
+        for (int i = 0; i < count; i++) {
+            var field = array[i];
+            if ((field.fieldFlags() & (ACC_STATIC | ACC_STRICT)) == ACC_STRICT) {
+                size++;
+            }
+        }
+        if (size == 0)
+            return UnsetField.EMPTY_ARRAY;
+        UnsetField[] ret = new UnsetField[size];
+        int j = 0;
+        for (int i = 0; i < count; i++) {
+            var field = array[i];
+            if ((field.fieldFlags() & (ACC_STATIC | ACC_STRICT)) == ACC_STRICT) {
+                ret[j++] = new UnsetField(AbstractPoolEntry.maybeClone(cpb, field.fieldName()),
+                        AbstractPoolEntry.maybeClone(cpb, field.fieldType()));
+            }
+        }
+        assert j == size : "toctou: " + j + " != " + size;
+        return ret;
+    }
+
+    // The captured information of unset fields, pool entries localized to class writing context
+    // avoid creating NAT until we need to write the fields to stack maps
+    record UnsetField(Utf8Entry name, Utf8Entry type) {
+        public static final UnsetField[] EMPTY_ARRAY = new UnsetField[0];
+
+        public static UnsetField[] copyArray(UnsetField[] incoming, int resultLen) {
+            assert resultLen <= incoming.length : resultLen + " > " + incoming.length;
+            return resultLen == 0 ? EMPTY_ARRAY : Arrays.copyOf(incoming, resultLen, UnsetField[].class);
+        }
+
+        public static boolean mismatches(UnsetField[] one, int sizeOne, UnsetField[] two, int sizeTwo) {
+            if (sizeOne != sizeTwo)
+                return true;
+            for (int i = 0; i < sizeOne; i++) {
```

Comment:
there could be a mismatch even if both arrays have the same length and same content but the order of the fields is different in each array, not sure if this could happen but we should consider it

Reply:
Originally this matches because I am only removing unsets. Now Dan told me that we need to consider adding more to unset fields if we have loops, so I will do a sort for the arrays to ensure they have some definite order.

Reply:
I have added more test cases for control flow merge with different DA results. In the new version, I sort the fields by their CP indices (their UTF8s are from the same CP).